vm/devices/support/fs/lxutil/src/windows/symlink.rs

```diff
@@ -0,0 +1,147 @@
+            ),
+            W32Ss::IO_REPARSE_TAG_MOUNT_POINT => (
+                &reparse.Anonymous.MountPointReparseBuffer.PathBuffer,
+                reparse
+                    .Anonymous
+                    .MountPointReparseBuffer
+                    .SubstituteNameOffset,
+                reparse
+                    .Anonymous
+                    .MountPointReparseBuffer
+                    .SubstituteNameLength,
+                0,
+            ),
+            _ => return Err(lx::Error::EIO),
+        }
+    };
+
+    // safety: The validity of the reparse buffer is provided by the caller. If the buffer is valid,
+    // the area pointed to by `buffer + offset` is a valid wstring of length `length`, and this operation is safe.
+    let substitute_name = unsafe {
+        UnicodeString::new(std::slice::from_raw_parts(
+            buffer.as_ptr().byte_offset(offset as _),
+            (length as usize) / size_of::<u16>(),
+        ))
+        .map_err(|_| lx::Error::EIO)?
+    };
+
+    Ok((substitute_name, flags))
+}
+
+/// Translates an absolute NT symlink target to an LX path.
+fn translate_absolute_target(
+    substitute_name: &UnicodeString,
+    state: &super::VolumeState,
+) -> lx::Result<UnicodeString> {
+    if state.options.sandbox || state.options.symlink_root.is_empty() {
+        // EPERM is the default return value if no callback is provided
+        return Err(lx::Error::EPERM);
+    }
+
+    // Convert from UTF-16 UNICODE_STRING to String
+    let name = substitute_name.as_slice();
+    if name.len() < 6 {
+        return Err(lx::Error::EIO);
+    }
+    let name = if name[name.len() - 1] == 0 {
+        &name[..name.len() - 1]
+    } else {
+        &name[..name.len()]
+    };
+    let name = match String::from_utf16(name) {
+        Ok(name) => name,
+        Err(_) => return Err(lx::Error::EIO),
+    };
+
+    // If the symlink does not start with \??\, it is malformed.
+    if !name.starts_with("\\??\\") {
+        return Err(lx::Error::EIO);
+    }
+
+    // Next must be a drive letter, a colon, and another separator.
+    // N.B. Mount-point junctions, which use a volume GUID style path, are not supported.
+    let (_, name) = name.split_at(4);
+    let mut name_as_chars = name.chars();
+    let drive_letter = match name_as_chars.next() {
+        Some(val) => val,
+        None => return Err(lx::Error::EIO),
+    };
+    if name_as_chars.next() != Some(':') || name_as_chars.next() != Some('\\') {
+        return Err(lx::Error::EIO);
+    };
+    let drive_letter = match drive_letter {
+        'a'..='z' => drive_letter,
+        'A'..='Z' => ((drive_letter as u8) - b'A' + b'a') as char,
+        _ => return Err(lx::Error::EIO),
+    };
+
+    let (_, name) = name.split_at(2);
+    let name = name.replace('\\', "/");
+    let target = format!("{}{}{}", &state.options.symlink_root, drive_letter, name);
+    let target = match U16CString::from_str(target) {
```

Comment:
It looks like all the callers will end up using this as a UTF-8 `String`. So it seems unnecessary to convert from a UTF-8 string to a null-terminated UTF-16 string to a counted UTF-16 string just to convert it back to UTF-8 in the callers.

Reply:
Maybe fixing this conflicts with your goal to keep the API the same. And so that's OK as a first step. But it would be good to add some TODOs for places where it makes sense to fix the API to fix our needs better.